nixos/modules/security/wrappers/wrapper.c

```diff
@@ -161,22 +162,25 @@ static int make_caps_ambient(const char *selfPath)
 
 int main(int argc, char * * argv)
 {
-    // I *think* it's safe to assume that a path from a symbolic link
-    // should safely fit within the PATH_MAX system limit. Though I'm
-    // not positive it's safe...
-    char selfPath[PATH_MAX];
-    int selfPathSize = readlink("/proc/self/exe", selfPath, sizeof(selfPath));
+    int selfExe = open("/proc/self/exe", O_PATH | O_CLOEXEC | O_NOFOLLOW);
+    assert(selfExe != -1);
+    struct stat st;
+    assert(fstat(selfExe, &st) != -1);
+    size_t selfPathCap = st.st_size + 1;
+    // Purposefully leak this, it's not going to be around for long...
+    char *selfPath = malloc(selfPathCap);
+    assert(selfPath);
+    int selfPathSize = readlinkat(selfExe, "", selfPath, selfPathCap);
 
     assert(selfPathSize > 0);
 
     // Assert we have room for the zero byte, this ensures the path
     // isn't being truncated because it's too big for the buffer.
     //
-    // A better way to handle this might be to use something like the
-    // whereami library (https://github.com/gpakosz/whereami) or a
-    // loop that resizes the buffer and re-reads the link if the
-    // contents are being truncated.
-    assert(selfPathSize < sizeof(selfPath));
+    // This probably isn't possible with readlinkat following fstat, but not hard to check...
+    assert(selfPathSize < selfPathCap);
```

Comment:
I agree, given that we are sizing `selfPathCap` correctly this code path is unlikely to be hit since we do not need to be concerned about truncation. I also think it is fine to leave the check in here.

However, would it be too much trouble to explain this a bit more in the comment?

Reply:
@ixmatus Sure, will do.